### PR TITLE
Layernorm code gen, simd and merged loops

### PR DESCRIPTION
**Pull request by AlexandreEichenberger**
When SIMD is on, will attempt to merge the elementary ONNX ops and generate SIMD code for it. The best way to explain the algorithm is maybe to give my Jupyter notebook test example.

Unoptimized functionality
```python
def layer_norm(x, a, scale, b):
    t_shape = x.shape
    t_rank = len(t_shape)
    mean = np.mean(x, axis=a, keepdims=True)
    d = np.subtract(x, mean)
    dd = np.multiply(d, d)
    print('dd', dd)
    var = np.mean(dd, axis=a, keepdims=True)
    print('var', var)
    var_eps = np.add(var, 1e-05)
    std_dev = np.sqrt(var_eps)
    inv_std_dev = np.divide(1.0, std_dev)
    normalized = np.multiply(d, inv_std_dev)
    normalized_scaled = np.multiply(normalized, scale)
    y = np.add(normalized_scaled, b)
    return (y, mean, inv_std_dev)

x = np.random.rand(4,8)
print('x', x)
(y, mean, inv_std_dev) = layer_norm(x, (-1), 1.0, 1.0)
print('mean', mean)
print('y', y)
```

Optimized functionality
```python
def layer_norm_simd2_v3(x, a, scale, b):
    (b1, b2) = (2, 4)
    t_rank = len(x.shape)
    assert t_rank==2, "work for rank 2 only"
    (a1, s2)  = x.shape
    mean = np.zeros((a1, 1))
    inv_std_dev = np.zeros((a1, 1))
    y = np.zeros((a1, s2))
    for i1 in range(0, a1, b1):
        # iterate over blocks of b1 values
        
        # MEAN(x), MEAN(x2)
        # iterate over a_block, s_block: parallel add
        r = np.zeros((b1, b2))
        r_square = np.zeros((b1, b2))
        for i2 in range(0, s2, b2): # Unroll B1, SIMD by B2, 
            xx = x[i1:i1+b1, i2:i2+b2]
            xxx = np.multiply(xx, xx)
            r = np.add(r, xx)
            r_square = np.add(r_square, xxx)
            
        # simd reduction; res B1 x 1
        # 2 B1 div
        mean_b = np.sum(r, axis=1, keepdims=True) # SIMD reduction to (B1x1) values.
        mean_b = np.divide(mean_b, s2) # (B2x1) values... so scalar is ok. 
        mean_square_b = np.sum(r_square, axis=1, keepdims=True) # Same.
        mean_square_b = np.divide(mean_square_b, s2) 

        #print('i1', i1, "mean", mean_b)
        
        # var = mean_square - mean**2; all compute here are on (B1x1): B1 mul, B1 add
        mean2_b = np.multiply(mean_b, mean_b) # B1 values, ok to do scalar
        var_b = np.subtract(mean_square_b, mean2_b)
        
        # ADD eps, sqrt, inverse
        # computations on B1x1, scalar ok: B1 add, B1 sqrt, B1 div
        var_eps_b = np.add(var_b, 1e-05)
        std_dev_b = np.sqrt(var_eps_b)
        inv_std_dev_b = np.divide(1.0, std_dev_b)
        #print('i1', i1, "inv std dev", inv_std_dev_b)

        # tot ops up to here (on B1x1): div: 3*B1, sqrt: B1, mul B1, add/sub 2 B1, sqrt B1: tot 8 B1

        # SIMD on entire S2 size
        for i2 in range(0, s2, b2): # Unroll B1, SIMD by B2, 
            x_b = x[i1:i1+b1, i2:i2+b2]
            d_b = np.subtract(x_b, mean_b) # broadcast of mean_b of 1 -> s2
            normalized_b = np.multiply(d_b, inv_std_dev_b)  # broadcast of mean_b of 1 -> s2
            normalized_scaled_b = np.multiply(normalized_b, scale) # assume here scale is scalar
            y_b = np.add(normalized_scaled_b, b) # assume here b is scalar

            # save values
            #mean[i1:i1+b1, i2:i2+b2] = mean_b
            #inv_std_dev[i1:i1+b1, i2:i2+b2] = inv_std_dev_b
            y[i1:i1+b1, i2:i2+b2] = y_b
        mean[i1:i1+b1, :] = mean_b
        inv_std_dev[i1:i1+b1, :] = inv_std_dev_b

    
    return (y, mean, inv_std_dev)

(y3_v3, mean3_v3, inv_std_dev3_v3) = layer_norm_simd2_v3(x, (-1), 1.0, 1.0)
print('y', y)
print('y3_v3', y3_v3)
print('mean', mean)
print('mean3_v3', mean3_v3)
print('inv std dev', inv_std_dev)
print('inv std dev3_v3', inv_std_dev3_v3)
```

### CODE REVIEW

**Comment by tungld** [comment]
```
# var = mean_square - mean**2; all compute here are on (B1x1): B1 mul, B1 add
mean2_b = np.multiply(mean_b, mean_b) # B1 values, ok to do scalar
var_b = np.subtract(mean_square_b, mean2_b)
```

Could you elaborate on how you derived this computation for `var_b`? How is it equivalent to `(x - mean_x)^2`

**Comment by tungld** [comment]
For complex operations, I often put the python script as comment at the end of the lowering file, it's helpful for debugging in the future.

**Comment by AlexandreEichenberger** [comment]
> For complex operations, I often put the python script as comment at the end of the lowering file, it's helpful for debugging in the future.

Good suggestion. Added the code and the proof that the approach is correct. It follows the algorithm used in the onnx builtin function (from third_party/onnx). I did spend some time then to convince myself that it was the same, back to high school algebra I guess. Now the proof is in the code base too.

**Comment by AlexandreEichenberger** [comment]
@tungld thanks for the review, I addressed all your comments.

File: src/Conversion/ONNXToKrnl/NN/Normalization.cpp [hunk 1]
```diff
@@ -383,10 +495,356 @@ struct ONNXLayerNormalizationOpLowering
...
+            xxx = np.multiply(xx, xx)
+            r = np.add(r, xx)
+            r_square = np.add(r_square, xxx)
+
+        # simd reduction; res B1 x 1
+        # 2 B1 div
+        mean_b = np.sum(r, axis=1, keepdims=True) # SIMD reduction to (B1x1)
+          # values. mean_b = np.divide(mean_b, s2) # (B2x1) values... so scalar
+          # is ok. mean_square_b = np.sum(r_square, axis=1, keepdims=True)
+          # Same.
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/NN/Normalization.cpp:808]: This part seems to be broken due to clang-format. The same for the below code for `normalized_scaled_b`. You can use `clang-format on/off` to keep the python code unmodified.

**Review by tungld — APPROVED** [review]
LGTM!

Thanks for the explanation! I am eager to see how much improvement we can get for roberta.

File: src/Conversion/ONNXToKrnl/NN/Normalization.cpp (hunk 1 above)
Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToKrnl/NN/Normalization.cpp:808]: Will do

**Comment by AlexandreEichenberger** [comment]
I don't have Roberta with layer norm, but a simple test example with the same dimensions shows the following on z16
```
Fused simd (new approach) 1.4e-4
Fused with disabled simd: 1.0e-3
Decompose:                1.7e-3
Decompose+NNPA:           1.4e-4
```
So it's about on par with Decompose NNPA. The outer loop is fully parallel, so CPU can be sped up by using parallelism. Will try once I get OMP working on Z.

**Pull request merged**